File: pypi/trackfw/fatal_error.py
"""fatal_error.py — global backstop for uncaught exceptions in trackfw.cli:main().

Defense in depth (REQ-2026-08-16-erro-nao-tratado-no-cli-node-vaza-stack-trace-
caminhos-absolutos-e-versao-do-runtime). The measured leak was in the Node CLI:
commands that catch internally (e.g. trackfw.integrations.command.run(), which
already prints "trackfw <kind> <action>: <error>" for IntegrationError/OSError/
ValueError) print clean; anything that escapes uncaught would otherwise let
Python's default excepthook dump the full traceback, including absolute
source-file paths and line numbers. This module is installed once, at the
single dispatch point in cli.py:main() — not per command — for the same
reason the Node fix lives in bin/trackfw and not in every command file: the
leak was per-command, and a per-entrypoint handler closes every future gap at
once.

TRACKFW_DEBUG=1 restores the full traceback — this must never blind someone
who genuinely needs to debug a crash.
"""

from __future__ import annotations

import os
import sys
import traceback


def report_fatal_error(error: BaseException, command: str | None = None) -> None:
    """Prints a clean, single-line error to stderr — or the full traceback when
    TRACKFW_DEBUG=1. Deliberately mirrors npm/src/lib/fatal-error.js:
    TRACKFW_DEBUG prints the traceback (which already includes the exception
    message as its own first/last line) instead of the traceback AND the
    clean message, to avoid printing the message twice.
    """
    if os.environ.get("TRACKFW_DEBUG") == "1":
        traceback.print_exc(file=sys.stderr)
        return
    message = str(error) or error.__class__.__name__
    prefix = f"trackfw {command}" if command else "trackfw"
    print(f"{prefix}: {message}", file=sys.stderr)

File: npm/src/lib/fatal-error.js
'use strict'

const fs = require('node:fs')

// fatal-error.js — global backstop for errors that escape every commander
// action (sync or async, since npm/bin/trackfw drives the program via
// parseAsync()). Installed ONCE at the entrypoint, never per-command:
// the real leak pattern (REQ-2026-08-16-erro-nao-tratado-no-cli-node-vaza-
// stack-trace-caminhos-absolutos-e-versao-do-runtime) was per-command —
// commands that catch internally (e.g. `roadmap move`) already print a
// clean message via console.error + process.exitCode, but any command that
// doesn't catch let Node's own uncaught-exception/unhandled-rejection
// printer dump the full stack, absolute install path and runtime version.
// A single handler at the entrypoint closes every one of those gaps at
// once instead of relying on every future command remembering a try/catch.
//
// TRACKFW_DEBUG=1 restores the full stack — this must never blind someone
// who genuinely needs to debug a crash.
//
// Writes with fs.writeSync(2, ...) — NOT process.stderr.write() — on
// purpose: installGlobalHandlers() below calls process.exit() right after
// this returns, and when stderr is a pipe (exactly what every spawnSync-
// based test in this repo uses, and what a piped CI log is) Node's stream
// writes are asynchronous; process.exit() can drop a still-pending write
// before the OS ever sees it, silently eating the message (failure mode
// #2 in the REQ: "engolir a mensagem" — worse than the leak this handler
// exists to fix). fs.writeSync is a raw, synchronous syscall — it cannot
// race process.exit().
function reportFatalError(err) {
  // TRACKFW_DEBUG=1: print the full stack (its first line already is
  // "Error: <message>", so this is the debug-only superset of the clean
  // path below, not a duplicate of it).
  if (process.env.TRACKFW_DEBUG === '1' && err && err.stack) {
    fs.writeSync(2, `${err.stack}\n`)
    return
  }
  const message = err && err.message !== undefined ? err.message : String(err)
  fs.writeSync(2, `Error: ${message}\n`)
}

// installGlobalHandlers — defense in depth beyond the parseAsync().catch()
// in bin/trackfw. Covers errors that reach the process outside that single
// promise chain (e.g. a rejection from a detached/floating promise started
// by a command, or a genuine uncaughtException). Per Node's own contract,
// once a listener is attached Node no longer prints its default trace and
// no longer auto-exits — the listener owns both the message and the exit.
// Both listeners call process.exit() immediately after reportFatalError()
// returns — safe here specifically because reportFatalError writes
// synchronously (see comment above); process.exitCode (which waits for the
// event loop to drain) would be wrong here regardless, since long-running
// commands (e.g. `serve`) hold the loop open and the process would hang
// instead of exiting on a genuinely fatal error.
function installGlobalHandlers() {
  process.on('unhandledRejection', (err) => {
    reportFatalError(err)
    process.exit(1)
  })
  process.on('uncaughtException', (err) => {
    reportFatalError(err)
    process.exit(1)
  })
}

module.exports = { reportFatalError, installGlobalHandlers }
